### PR TITLE
Implementing tom2

### PR DESCRIPTION
**Pull request by jfly**
Added a lastAxisRestriction to min2phase's Search in addition to the existing firstAxisRestriction.
This lets us generate FMC scrambles that
start and end with the same turns because we can make sure that the
prefix and suffix won't cancel with the moves of the original scramble.

Here's a quick example of it working:

```
~/gitting/tnoodle @kaladin> ./jython scrambles/test/ThreeByThreeCubeFewestMovesTest.jy
...
Solution to F' (solution may not start with F axis and may not end with B axis): U2 F B' R2 F R2 F' B U2
Solution to D' (solution may not start with D axis and may not end with U axis): R2 U D' F2 D F2 U' D R2
Solution to D' (solution may not start with D axis and may not end with R axis): R2 L2 U2 D2 R2 L2 U2 D'
Solution to D' (solution may not start with D axis and may not end with F axis): R2 L2 U2 D2 R2 L2 U2 D'
Solution to D' (solution may not start with D axis and may not end with D axis): R2 U D' F2 D F2 U' D R2
Solution to D' (solution may not start with D axis and may not end with L axis): R2 L2 U2 D2 R2 L2 U2 D'
Solution to D' (solution may not start with D axis and may not end with B axis): R2 L2 U2 D2 R2 L2 U2 D'
Solution to L' (solution may not start with L axis and may not end with U axis): F2 B2 R2 L2 F2 B2 R2 L'
Solution to L' (solution may not start with L axis and may not end with R axis): F2 R L' U2 L U2 R' L F2
Solution to L' (solution may not start with L axis and may not end with F axis): F2 B2 R2 L2 F2 B2 R2 L'
Solution to L' (solution may not start with L axis and may not end with D axis): F2 B2 R2 L2 F2 B2 R2 L'
Solution to L' (solution may not start with L axis and may not end with L axis): F2 R L' U2 L U2 R' L F2
Solution to L' (solution may not start with L axis and may not end with B axis): F2 B2 R2 L2 F2 B2 R2 L'
Solution to B' (solution may not start with B axis and may not end with U axis): U2 D2 F2 B2 U2 D2 F2 B'
Solution to B' (solution may not start with B axis and may not end with R axis): U2 D2 F2 B2 U2 D2 F2 B'
Solution to B' (solution may not start with B axis and may not end with F axis): U2 F B' R2 B R2 F' B U2
Solution to B' (solution may not start with B axis and may not end with D axis): U2 D2 F2 B2 U2 D2 F2 B'
Solution to B' (solution may not start with B axis and may not end with L axis): U2 D2 F2 B2 U2 D2 F2 B'
Solution to B' (solution may not start with B axis and may not end with B axis): U2 F B' R2 B R2 F' B U2
25 move 333fm scramble: R' U' F L2 U' R2 F2 L2 R2 D R2 B F D' L R2 D L2 F2 U L2 U2 R' U' F
26 move 333fm scramble: R' U' F U F2 U' F2 U' R2 B2 U F2 D' U2 B' U2 F' D2 L R2 F' L' D R' U' F
27 move 333fm scramble: R' U' F R B2 L U2 R2 D2 R' F2 D2 B2 D2 U' F D2 F2 L2 R2 F' L2 R U R' U' F
27 move 333fm scramble: R' U' F U2 F2 U' B2 F2 R2 B2 U2 R2 D2 L F' L' R' D2 U L2 R U' B2 D' R' U' F
25 move 333fm scramble: R' U' F L2 D U R2 B2 R2 D' L2 F' U2 F U' L D2 U2 F U F U2 R' U' F
26 move 333fm scramble: R' U' F D' B2 U' R2 D L2 D2 R2 D' B' D' B' L' B R' F' U2 F2 D' F R' U' F
27 move 333fm scramble: R' U' F R2 F2 D2 L2 B' D2 U2 R2 F U' R F2 R2 U F' L' U R2 U' R D' R' U' F
25 move 333fm scramble: R' U' F L2 R2 U' F2 R2 U F2 U F' D U2 L' B2 D2 B2 R U R B2 R' U' F
26 move 333fm scramble: R' U' F U' R2 U2 L2 D2 B2 U' B2 F' U' F L2 B' U' R' B2 F L' B U2 R' U' F
26 move 333fm scramble: R' U' F R2 B2 U L2 D' F2 U' R2 U R B D2 R D' U R D R2 B' F R' U' F
```

Also:

image

### CODE REVIEW

File: scrambles/src/puzzle/ThreeByThreeCubeFewestMovesPuzzle.java [hunk 1]
```diff
@@ -17,7 +17,57 @@ public ThreeByThreeCubeFewestMovesPuzzle() {
...
+        // The prefix/suffix will be easily recognizable and memorable and ideally finger-trick
+        // friendly (if it matters).
+        //
+        // Someone wanting to practice speed FMC (hehe) can make up their own scrambles just
+        // by manually doing the prefix/suffix thing on any existing scrambler.
+        //
+        // It does not perturb the uniformity of the random position selection.
+        //
+        // It's simple enough that it is less likely to suffer from subtle defects due to the
+        // perturbation of the two-phase search (unlikely those these may be).
```
Comment (lgarron) [inline scrambles/src/puzzle/ThreeByThreeCubeFewestMovesPuzzle.java]: "unlikely *as* those may be"?

Comment (lgarron) [inline scrambles/src/puzzle/ThreeByThreeCubeFewestMovesPuzzle.java]: Also, it doesn't really make sense to say whether a specific scramble was perturbed. The distribution is a *little bit* perturbed for *all scrambles*.

Comment (lgarron) [inline scrambles/src/puzzle/ThreeByThreeCubeFewestMovesPuzzle.java]: To take a timely but politically charged example, it's the difference between affirmative action and privilege.

Comment (jfly) [inline scrambles/src/puzzle/ThreeByThreeCubeFewestMovesPuzzle.java]: Thanks for taking a look, Lucas. Could you submit a PR with updated language?

**Pull request merged**